### ISSUE TITLE
What happens if we don't use cross-validation (CV) and just use trainset and test-set within our forecasting on tail of time data?

### ISSUE THREAD

**Issue by clevilll** [skforecast/skforecast#887]
@JoaquinAmatRodrigo I have a question that I did not use cross-validation (CV) like you have done in _[Data Partition](https://cienciadedatos.net/documentos/py60-probabilistic-forecasting-prediction-intervals-multi-step-forecasting#Data-partition)_ section in recent notebook. I just split dat into train and test-set (without validation-set) and I plotted for your consideration.
![image](https://github.com/user-attachments/assets/0aecfa76-3f95-4135-a884-f101eb6bf9b6)

1. Do you think is it critical and used ML-based  regression models' learning could be suffer from over/under-fiting when i did not consider CV-set?
2. base on the picture do you think I _damaged_ nature of time-data in the plot after pre-processing stage (de-noise filter, fill missing sequences, detect and replaced global outliers)? in the plot as legend shows cleaned data divided into train-set and test-set and plot over raw time data.

I also did not also use GridSerachCV within my pipeline also due to save runtime! I have lost of `df`s samples I need to apply the designed pipeline. Can  you kindly comment on these Qs separately

**Comment by MarcosBelver** [comment]
Hi @clevilll 

For question 1, in my experience, what you are doing is evaluating you model's performance in only a small portion of the data. In order to have more robust idea about this performance, CV should be used in some cases (if it is possible, of course). In your case, your test data is quite long (I suppose because your forecast horizon is quite large, approximately 2000 time steps). If you need to predict so many steps in the future, is true that in your case you won't have too much data to create the splits (probably not more than 3 folds, since you need some data to train with). Anyway, I would encourage you to implement the CV with 3 folds, since in this way are you going to have a more realistic idea about the model performance over time.

In terms of over/under-fitting, this concept is specially important when you decide to fine-tune your model hyperparameters. In this case I think that is even more important to do that using a small CV process in order to not bias results.



About question 2, I do not fully understand. In some cases, it is needed to apply pre-processing steps as you mention and in most of the cases is totally justified to remove noise and fill missing points. In some of my projects (for example, related to price prediction) when I observe anomalies I try to correct them, since it is not the expected behavior. Anyway, everything depends on the context. There are other situation where forecasters do not recommend to threat outliers and model them through the model. 


Hopes this can help you

**Comment by clevilll** [comment]
@MarcosBelver, thanks for your reply.

I'm experimenting with conformal prediction over high-frequent time data using the following forest-based regression models for an in-sample forecasting task. The size of univariate (1D) time-series data is `(8640, 1)`, meaning a single column value over time for 8460 time steps.  

> **short note:** 
>
> 8460 means $30×288$  (every day has 288 observations) -- data collected with **epochs of 5 mins** which leads $12×24$ = 288 observations\data per day 

So, what is the best setup for my experiment? [2w:1w:1w] Is ok? Assuming having 4weeks time data like:


```python
# Configuration
# ==============================================================================
steps = 2017 #7*288
lags = 288
name_columns = "avgcpu"
```
When I was checking this point: avoid potential overfitting, I checked this post [Overfitting - how to detect it and reduce it?](https://datascience.stackexchange.com/q/46646/69263) for R language that led me to this post [how to choose a good validation set](https://www.fast.ai/2017/11/13/validation-sets/). Please let me know what you suggest for my time data.

**Comment by MarcosBelver** [comment]
Hi @clevilll 

I think everything depends on the validation set (your forecast horizon, actually) because, in my opinion, it is what is going to indicate the configuration of your experiment to a great extent. 

In this case you want to predict always one week ahead. In that case, I think you should create a CV-3 folds or something like that (if you want to do fine-tuning, then I would reconsider this, leaving maybe a one test set aside to evaluate performance on unseen data). 

If you use skforecast CV method, it will do it automatically for you. You will obtain something like this (fixing your fh=2017):

**CV-1:**
Train set: from 0 to 2588
Val set: from 2589 to 4605

**CV-2:**
Train set: from 0 to 4605
Val set: from 4606 to 6622

**CV-3:**
Train set: from 0 to 6622
Val set: from 6623 to 8639

This is orientative and representing the indexes. There are many other alternatives already implemented in skforecast.

Check this:

https://skforecast.org/0.13.0/user_guides/backtesting

In your case, I would consider the simple approach (the one I used above) and the approach based on fixed window (for example, training with a size of 1 week and forecasting the next one. This approach could be interesting in your case I think).

![image](https://github.com/user-attachments/assets/14f74e26-7035-434c-932d-5b33f25b3467)


But remember, this is for testing purposes. With 3-cv fold approach you will obtain a more robust evaluation through the historical data. If you want to do fine-tuning, I would reconsider this a bit, perhaps leaving the last week as unseen data for not overfitting the model. In that case, a 3-cv fold approach could not be suitable given the amount of remaining data.   

Let me know if it is clearer now.


Regards,

**Comment by clevilll** [comment]
Thanks for your recommendations.

I will definitely use the different backtesting sk proposed.

Let me see if I understand you correctly, so you say instead of doing _[Data Partition](https://cienciadedatos.net/documentos/py60-probabilistic-forecasting-prediction-intervals-multi-step-forecasting#Data-partition)_ section in a recent notebook means slicing data into train-set, val-set, and test-set (unseen) and we do  3-cv fold  as you proposed to the apply  `ForecasterRecursive()` (recommended by @JoaquinAmatRodrigo in this [issue](https://github.com/cerlymarco/linear-tree/issues/43#issuecomment-2481647437))
 or
we can directly use _backtesting with fixed traind size_ means `backtesting_forecaster(refit=True, fixed_train_size=True)` and configure (fixing my future-horizon=7*288=2016) which automatically does the job 3-CV, right?

Then I will have something like :
```none
Number of observations used for initial training: ...
Number of observations used for backtesting: ..
    Number of folds: 3
    Number skipped folds: 0 
    Number of steps per fold: ..
....

Fold: 0
    Training:   1991-07-01 00:00:00 -- 2002-01-01 00:00:00  (n=127)
    Validation: 2002-07-01 00:00:00 -- 2003-04-01 00:00:00  (n=10)
Fold: 1
    Training:   1991-07-01 00:00:00 -- 2002-11-01 00:00:00  (n=137)
    Validation: 2003-05-01 00:00:00 -- 2004-02-01 00:00:00  (n=10)
Fold: 2
    Training:   1991-07-01 00:00:00 -- 2003-09-01 00:00:00  (n=147)
    Validation: 2004-03-01 00:00:00 -- 2004-12-01 00:00:00  (n=10)
```

**Comment by JoaquinAmatRodrigo** [comment]
Hi everyone! 👋

It’s been a while since the last activity on this issue, so I’m going to close it to help us keep our backlog manageable and focused on active tasks.

If you are still experiencing this problem or believe this feature is still highly relevant for the latest version, please feel free to leave a comment or reopen the issue. Thank you for your contribution!
